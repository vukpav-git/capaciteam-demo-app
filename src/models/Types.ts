export interface IBillDataRow {
  bill: {
    billNo: string;
    billType: string;
    status: string;
    sponsors: [
      {
        sponsor: {
          as: {
            showAs: string;
          };
          by: {
            showAs: string;
          };
        };
      },
    ];
    favorites?: boolean;
  };
}

export interface IColumnHead {
  name: string;
  align: "left" | "right";
}

export type TLanguage = "en" | "ga";

export type TBillType = "All" | "Public" | "Private";

export enum BillTypes {
  All = "All",
  Public = "Public",
  Private = "Private",
}
