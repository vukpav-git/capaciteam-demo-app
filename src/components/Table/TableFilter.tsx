import { Box, FormControl, InputLabel, MenuItem, Select } from "@mui/material";
import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";

import { BillTypes, type TBillType } from "../../models/Types";
import {
  FILTER_BILLS_TYPE,
  FILTER_BILLS_TYPE_ENDED,
  FILTER_BILLS_TYPE_STARTED,
} from "../../store/actions/ActionTypes";

const TableFilter = () => {
  const [billType, setBillType] = useState<TBillType>(BillTypes.All);
  const dispatch = useDispatch();

  const favoritesFiltered = useSelector(
    (state: any) => state.favoritesFiltered,
  );

  useEffect(() => {
    // reset filter on tab change
    setBillType(BillTypes.All);
  }, [favoritesFiltered]);

  const handleChange = (typeVal: TBillType) => {
    dispatch({ type: FILTER_BILLS_TYPE_STARTED });
    setBillType(typeVal);

    dispatch({ type: FILTER_BILLS_TYPE, payload: typeVal });
    dispatch({ type: FILTER_BILLS_TYPE_ENDED });
  };

  return (
    <Box marginBottom="12px" display="flex" justifyContent="flex-end">
      <FormControl
        data-testid="form-control-id"
        variant="standard"
        sx={{ m: 1, minWidth: "242px" }}
      >
        <InputLabel sx={{ fontSize: 18 }}>Filter by "Bill type"</InputLabel>
        <Select
          data-testid="TableFilter-select"
          labelId="TableFilter-label"
          value={billType}
          onChange={(e: any) => handleChange(e.target.value)}
          label="Bill Type"
        >
          <MenuItem value="All">All</MenuItem>
          <MenuItem value="Public">Public</MenuItem>
          <MenuItem value="Private">Private</MenuItem>
        </Select>
      </FormControl>
    </Box>
  );
};

export default TableFilter;
