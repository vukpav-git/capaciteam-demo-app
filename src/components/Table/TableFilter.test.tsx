import { fireEvent, render, screen, within } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

import { BillTypes } from "../../models/Types";
import TableFilter from "./TableFilter";

const mockStore = configureStore([]);
// Mock the initial state...
const store = mockStore({
  favoritesFiltered: false,
});

describe("Table Filter tests", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <TableFilter />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly and dispatches the correct actions on selection change", () => {
    render(
      <Provider store={store}>
        <TableFilter />
      </Provider>,
    );

    const selectCompoEl = screen.getByTestId("TableFilter-select");

    const button = within(selectCompoEl).getByRole("combobox");
    fireEvent.mouseDown(button);

    const listbox = within(screen.getByRole("presentation")).getByRole(
      "listbox",
    );

    const options = within(listbox).getAllByRole("option");

    fireEvent.click(options[2]);

    const actions = store.getActions();
    expect(actions).toEqual([
      { type: "bill/filterBillsByTypeStarted" },
      { type: "bill/filterBillsByType", payload: BillTypes.Private },
      { type: "bill/filterBillsByTypeEnded" },
    ]);
  });
});
