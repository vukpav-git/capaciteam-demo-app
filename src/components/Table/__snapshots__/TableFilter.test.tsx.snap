// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Filter tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-18tbono"
  >
    <div
      class="MuiFormControl-root css-1cp56a0-MuiFormControl-root"
      data-testid="form-control-id"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1p6z1ep-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
      >
        Filter by "Bill type"
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl  css-n70jm4-MuiInputBase-root-MuiInput-root-MuiSelect-root"
        data-testid="TableFilter-select"
      >
        <div
          aria-controls=":r0:"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="TableFilter-label"
          class="MuiSelect-select MuiSelect-standard MuiInputBase-input MuiInput-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
          role="combobox"
          tabindex="0"
        >
          All
        </div>
        <input
          aria-hidden="true"
          aria-invalid="false"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          tabindex="-1"
          value="All"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
